phpBB/phpbb/textformatter/s9e/renderer.php

```diff
@@ -0,0 +1,285 @@
+			}
+
+			if (isset($renderer_data['censor']))
+			{
+				$censor = $renderer_data['censor'];
+			}
+		}
+
+		if (!isset($renderer))
+		{
+			$objects  = $factory->regenerate();
+			$renderer = $objects['renderer'];
+		}
+
+		if (isset($censor))
+		{
+			$this->censor = $censor;
+		}
+
+		$this->renderer = $renderer;
+	}
+
+	/**
+	* Automatically set the smilies path based on config
+	*
+	* @param  \phpbb\config\config $config
+	* @param  \phpbb\path_helper   $path_helper
+	* @return null
+	*/
+	public function configure_smilies_path(\phpbb\config\config $config, \phpbb\path_helper $path_helper)
+	{
+		/**
+		* @see smiley_text()
+		*/
+		$root_path = (defined('PHPBB_USE_BOARD_URL_PATH') && PHPBB_USE_BOARD_URL_PATH) ? generate_board_url() . '/' : $path_helper->get_web_root_path();
+
+		$this->set_smilies_path($root_path . $config['smilies_path']);
+	}
+
+	/**
+	* Configure this renderer as per the user's settings
+	*
+	* Should set the locale as well as the viewcensor/viewflash/viewimg/viewsmilies options.
+	*
+	* @param  \phpbb\user          $user
+	* @param  \phpbb\config\config $config
+	* @param  \phpbb\auth\auth     $auth
+	* @return null
+	*/
+	public function configure_user(\phpbb\user $user, \phpbb\config\config $config, \phpbb\auth\auth $auth)
+	{
+		$censor = $user->optionget('viewcensors') || !$config['allow_nocensors'] || !$auth->acl_get('u_chgcensors');
+
+		$this->set_viewcensors($censor);
+		$this->set_viewflash($user->optionget('viewflash'));
+		$this->set_viewimg($user->optionget('viewimg'));
+		$this->set_viewsmilies($user->optionget('viewsmilies'));
+
+		// Set the stylesheet parameters
+		foreach (array_keys($this->renderer->getParameters()) as $param_name)
+		{
+			if (substr($param_name, 0, 2) === 'L_')
+			{
+				// L_FOO is set to $user->lang('FOO')
+				$this->renderer->setParameter($param_name, $user->lang(substr($param_name, 2)));
+			}
+		}
+
+		// Set the style id
+		$this->renderer->setParameter('STYLE_ID', $user->style['style_id']);
+	}
+
+	/**
+	* Return the instance of s9e\TextFormatter\Renderer used by this object
+	*
+	* @return s9e\TextFormatter\Renderer
+	*/
+	public function get_renderer()
+	{
+		return $this->renderer;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function get_viewcensors()
+	{
+		return $this->viewcensors;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function get_viewflash()
+	{
+		return $this->viewflash;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function get_viewimg()
+	{
+		return $this->viewimg;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function get_viewsmilies()
+	{
+		return $this->viewsmilies;
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function render($text)
+	{
+		if (isset($this->censor) && $this->viewcensors)
+		{
+			// NOTE: censorHtml() is XML-safe
+			$text = $this->censor->censorHtml($text, true);
+		}
+
+		$html = $this->renderer->render($text);
+
+		/**
+		* @see bbcode::bbcode_second_pass_code()
+		*/
+		$html = preg_replace_callback(
+			'#(<code>)(.*?)(</code>)#is',
```

Comment:
the text formatter should be generic and not tied to a single style. So yes I think you should handle the case where the code tag include other data such as a class